File: contests/lc/2943.MaximizeAreaofSquareHoleinGrid/solution.py
from typing import List


class Solution:
    def maximizeSquareHoleArea(
        self, n: int, m: int, hBars: List[int], vBars: List[int]
    ) -> int:
        max_h = self._max_consecutive_span(hBars)
        max_v = self._max_consecutive_span(vBars)
        side = min(max_h, max_v)
        return side * side

    @staticmethod
    def _max_consecutive_span(bars: List[int]) -> int:
        if not bars:
            return 1
        bars.sort()
        longest = 1
        current = 1
        for i in range(1, len(bars)):
            if bars[i] == bars[i - 1] + 1:
                current += 1
            else:
                if current > longest:
                    longest = current
                current = 1
        if current > longest:
            longest = current
        return longest + 1

File: contests/lc/2943.MaximizeAreaofSquareHoleinGrid/solution.go
package main

import "sort"

func maximizeSquareHoleArea(n int, m int, hBars []int, vBars []int) int {
	maxH := maxConsecutiveSpan(hBars)
	maxV := maxConsecutiveSpan(vBars)
	if maxH < maxV {
		return maxH * maxH
	}
	return maxV * maxV
}

func maxConsecutiveSpan(bars []int) int {
	if len(bars) == 0 {
		return 1
	}
	sort.Ints(bars)
	longest := 1
	current := 1
	for i := 1; i < len(bars); i++ {
		if bars[i] == bars[i-1]+1 {
			current++
		} else {
			if current > longest {
				longest = current
			}
			current = 1
		}
	}
	if current > longest {
		longest = current
	}
	return longest + 1
}
